### PR TITLE
[VB Compiler Scanner] Unifies TryScanToken & ScanTokenFullWidth

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
[VB Compiler Scanner] Unifies two methods that had similar content to use a common implementation.

Reduces the size of both of the source code and resulting IL

```
                       Old |  New | Rel
                     ------+------+------
      TryScanToken:-  1746 |   42 |   34
ScanTokenFullWidth:-  1972 |   15 |   10
   ScanTokenCommon:-       | 2035 | 1665
                     ------+------+------
             Total:-  3718 | 2092 | 1709

+ Minor Whitespace changes typical on the end of lines.

### CODE REVIEW

**Comment by AdamSpeight2008** [comment]
@gafter @AlekseyTs @jaredpar

Ready for review? Is it small enough? :smirk:

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 '-----------------------------------------------------------------------------
-' Contains the definition of the Scanner, which produces tokens from text 
+' Contains the definition of the Scanner, which produces tokens from text
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: What is causing changes like this?

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb [hunk 2]
```diff
@@ -1199,247 +1171,47 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
...
-                    End If
-
-            End Select
-
-            If IsIdentifierStartCharacter(ch) Then
-                Return ScanIdentifierOrKeyword(precedingTrivia)
-            End If
+        ' at this point it is very likely that we are located at the beginning of a token
+        Private Function TryScanToken(precedingTrivia As SyntaxList(Of VisualBasicSyntaxNode)) As SyntaxToken
+            If CanGet Then Return ScanTokenCommon(precedingTrivia, Peek(), False)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Please convert this to block-If.

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb [hunk 3]
```diff
@@ -1199,247 +1171,47 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
...
+            Return MakeEofToken(precedingTrivia)
+        End Function
 
-            Debug.Assert(Not IsNewLine(ch))
-            Debug.Assert(Not IsDoubleQuote(ch))
+        Private Function ScanTokenFullWidth(precedingTrivia As SyntaxList(Of VisualBasicSyntaxNode), ch As Char) As SyntaxToken
+            Return ScanTokenCommon(precedingTrivia, ch, True)
+         End Function
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Alignment of ```End Function``` is off.

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb [hunk 4]
```diff
@@ -1199,247 +1171,47 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
...
-
             If IsIdentifierStartCharacter(ch) Then
                 Return ScanIdentifierOrKeyword(precedingTrivia)
             End If
-
             Debug.Assert(Not IsNewLine(ch))
-
+            If fullWidth Then
+                Debug.Assert(Not IsDoubleQuote(ch))
+               Return nothing
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: ```Nothing```?

**Comment by AlekseyTs** [comment]
@AdamSpeight2008 Please address the feedback mentioned above. In order to reduce friction for future PRs, consider reverting whitespace changes before submitting a PR, especially outside of code being modified/added.

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb (hunk 4 above)
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: That is pre-existing within the codebase.
 [Scanner.vb#L1232](https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb#L1232)

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb (hunk 1 above)
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Extension [Trim line ends on save](https://visualstudiogallery.msdn.microsoft.com/5120dd1c-1643-4f0d-ae2b-c61532df1951)

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Consider disabling it, it creates unnecessary noise.

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Sorry for not being clear. Please start it with capital ```N```.

**Comment by AlekseyTs** [comment]
LGTM. @gafter, @VSadov, @jaredpar, @agocke Please review.

**Comment by VSadov** [comment]
The reason why there is a separate implementation for TryScanToken with hardcoded fullWidth=false is because the case where sources do not contain fullWidth characters is vastly more common. 
The idea is to make the most common code path to contain as little logic and branches as possible. And if we do have a case of a fullWidth character, there is a method specialized just for that. 

Indeed, that results in more code, but it also simplifies the most common codepath thus potentially making it faster. The realities here is that the end user will not care as much about how many IL instructions we have emitted, as he will care about how fast we are scanning.

Basically of two metrics - 1) text and IL size and 2) scanning throughput, the second is clearly more important. The change definitely improves 1, but are we sure it is not sacrificing 2?

Do we have any kind of assurance/measurements that scanning throughput did not regress after this change?

**Comment by VSadov** [comment]
We used to have some benchmarking projects but they relied on various internal tooling and I think they did not make it to the open source, so I am not sure there is an easy solution here.

One way in which this can be at least partially solved is perhaps adding a parser testcase that basically builds a big piece of VB code and parses it in a loop (to reduce noise the test should take hundreds of milliseconds to run) and prints the time in milliseconds it took to run the test. To not slowdown regular testing the test could just run one iteration to make sure the test is still valid - i.e. parsing does not produce errors and such. 

Since we are going to measure parser only, the sample only needs to be correct grammatically. For example repeated copy of Scanner class from Scanner.vb should work.

Could you make such test? It seems it would be very handy.

**Comment by AdamSpeight2008** [comment]
@VSadov
Unfortunately I'm not that skilled in regards to Unit Test. Another PR (#2339) was amended by someone else with the Unit Test.

We would also require a de-facto piece of code, to use a the standard. 
My initial thoughts was to use some version of `scanner.vb` as it's ~2500 IDE lines..
It utilise most of the language features (excluding interpolation).

I'm kinda curious what the JIT would do with `ScanTokenCommon` method.

**Comment by VSadov** [comment]
I have written a quick small test that runs parser with and without the change and there was no noticeable difference. It seems that the change does not have measurable impact on end-to-end parsing performance.

We still need some tests that could be used as benchmarks for changes like this. I will enter an issue to track that.

**Comment by VSadov** [comment]
LGTM

**Comment by AdamSpeight2008** [comment]
@VSadov Way ahead of you. (See #3602)

**Comment by agocke** [comment]
I've got both of you covered: #3567.
